olp-cpp-sdk-core/src/http/winhttp/NetworkWinHttp.cpp

```diff
@@ -0,0 +1,1006 @@
+      0);
+
+  if (chars_required == 0) {
+    // error: could not convert input string
+    return false;
+  }
+
+  out.resize(chars_required);
+  // Perform actual conversion from multi-byte to wide char
+  const auto conversion_result =
+      MultiByteToWideChar(CP_ACP, MB_PRECOMPOSED, in.c_str(),
+                          -1,  // denotes null-terminated string
+                          &out.front(), out.size());
+
+  // Should not happen as 1st call have succeeded.
+  return conversion_result != 0;
+}
+
+std::wstring ProxyString(const olp::http::NetworkProxySettings& proxy) {
+  using namespace olp::http;
+  std::wostringstream proxy_string_stream;
+
+  switch (proxy.GetType()) {
+    case NetworkProxySettings::Type::NONE:
+      proxy_string_stream << "http://";
+      break;
+    case NetworkProxySettings::Type::SOCKS4:
+      proxy_string_stream << "socks4://";
+      break;
+    case NetworkProxySettings::Type::SOCKS5:
+      proxy_string_stream << "socks5://";
+      break;
+    case NetworkProxySettings::Type::SOCKS4A:
+      proxy_string_stream << "socks4a://";
+      break;
+    case NetworkProxySettings::Type::SOCKS5_HOSTNAME:
+      proxy_string_stream << "socks5h://";
+      break;
+    default:
+      proxy_string_stream << "http://";
+  }
+
+  proxy_string_stream << std::wstring(proxy.GetUsername().begin(),
+                                      proxy.GetUsername().end());
+  proxy_string_stream << ':';
+  proxy_string_stream << proxy.GetPort();
+
+  return proxy_string_stream.str();
+}
+
+}  // namespace
+
+namespace olp {
+namespace http {
+
+#define LOGTAG "WinHttp"
+
+NetworkWinHttp::NetworkWinHttp()
+    : http_session_(NULL),
+      thread_(INVALID_HANDLE_VALUE),
+      event_(INVALID_HANDLE_VALUE) {
+  request_id_counter_.store(
+      static_cast<RequestId>(RequestIdConstants::RequestIdMin));
+
+  http_session_ = WinHttpOpen(L"OLP Http Client", WINHTTP_ACCESS_TYPE_NO_PROXY,
+                              WINHTTP_NO_PROXY_NAME, WINHTTP_NO_PROXY_BYPASS,
+                              WINHTTP_FLAG_ASYNC);
+
+  if (!http_session_) {
+    EDGE_SDK_LOG_ERROR(LOGTAG, "WinHttpOpen failed " << GetLastError());
+    return;
+  }
+
+  WinHttpSetStatusCallback(
+      http_session_, (WINHTTP_STATUS_CALLBACK)&NetworkWinHttp::RequestCallback,
+      WINHTTP_CALLBACK_FLAG_ALL_COMPLETIONS | WINHTTP_CALLBACK_FLAG_HANDLES, 0);
+
+  event_ = CreateEvent(NULL, TRUE, FALSE, NULL);
+
+  // Store the memory tracking state during initialization so that it can be
+  // used by the thread.
+  thread_ = CreateThread(NULL, 0, NetworkWinHttp::Run, this, 0, NULL);
+  SetThreadPriority(thread_, THREAD_PRIORITY_ABOVE_NORMAL);
+}
+
+NetworkWinHttp::~NetworkWinHttp() {
+  std::vector<std::shared_ptr<ResultData>> pendingResults;
+  {
+    std::unique_lock<std::recursive_mutex> lock(mutex_);
+    std::vector<std::shared_ptr<RequestData>> requestsToCancel;
+    for (auto& request : http_requests_) {
```

Comment:
```suggestion
    for (const auto& request : http_requests_) {
```
It doesn't change anything - so better to make const to eliminate invalid assignment to `request`